### PR TITLE
Add Solana

### PR DESCRIPTION
**Pull request by peachbits**
---
- To see the specific tasks where the Asana app for GitHub is being used, see below:
  - https://app.asana.com/0/0/1201425242712354

### CODE REVIEW

File: src/solana/solanaEngine.js [hunk 1]
```diff
@@ -0,0 +1,390 @@
...
+  settings: SolanaSettings
+
+  constructor(
+    currencyPlugin: SolanaPlugin,
+    walletInfo: EdgeWalletInfo,
+    opts: any, // EdgeCurrencyEngineOptions
+    fetchCors: EdgeFetchFunction
+  ) {
+    super(currencyPlugin, walletInfo, opts)
+    this.keypair = currencyPlugin.createKeyPair(
```
Comment (paullinator) [inline src/solana/solanaEngine.js:50]: This should not happen on every engine creation. createKeyPair uses slow cryptography. This should only be done when needed to sign a transaction. The public key should be created at the time you import or create the private key and be returned in the keys object.

File: src/solana/solanaEngine.js [hunk 2]
```diff
@@ -0,0 +1,390 @@
...
+      this.updateBalance(this.chainCode, balance.value.toString())
+    } catch (e) {
+      // fetching of account balances for uninitiated accounts throws error
+      if (
+        this.tokenCheckTransactionsStatus[this.chainCode] === 1 &&
+        this.transactionList[this.chainCode].length === 0
+      ) {
+        this.updateBalance(this.chainCode, '0')
+      }
+      this.error(`Error checking ${this.chainCode} address balance`, e)
```
Comment (paullinator) [inline src/solana/solanaEngine.js:119]: This isn't really an error if the account is uninitiated. Can you differentiate a true error and only log that?

File: src/solana/solanaEngine.js [hunk 3]
```diff
@@ -0,0 +1,390 @@
...
+            commitment: this.settings.commitment
+          }
+        ]
+        const response: RpcSignatureForAddress[] = await this.fetchRpc(
+          'getSignaturesForAddress',
+          params
+        )
+
+        // remove failed transactions from list
+        txids = txids.concat(
```
Comment (paullinator) [inline src/solana/solanaEngine.js:201]: Don't remove failed transactions if they incurred a fee to the user's wallet. Just like ETH, those are valid transactions, and the fee should still be shown. Set nativeamount to 0 and add it.

File: src/solana/solanaEngine.js [hunk 4]
```diff
@@ -0,0 +1,390 @@
...
+      return
+    }
+
+    if (txids.length === 0) {
+      this.tokenCheckTransactionsStatus[this.chainCode] = 1
+      this.updateOnAddressesChecked()
+      return
+    }
+
+    let newestTxid = 0
```
Comment (paullinator) [inline src/solana/solanaEngine.js:219]: Should be called `newestTxIndex`. It's not a txid

File: src/solana/solanaEngine.js [hunk 5]
```diff
@@ -0,0 +1,390 @@
...
+      } catch (e) {
+        // Note the oldest failed tx query so we try again next loop
+        newestTxid = i
+      }
+    }
+
+    // Don't update the newestTxid if the txids array length is length 1 or the oldest failed query is the end of the array.
+    // The previously saved value is the best to use in those cases
+    if (txids.length > 1 && txids.length > newestTxid + 1)
+      this.otherData.newestTxid = txids[newestTxid + 1].signature
```
Comment (paullinator) [inline src/solana/solanaEngine.js:244]: `newestTxid` has an inconsistent type. sometimes a number, sometimes a string. make it consistent.

**Review by paullinator — CHANGES_REQUESTED** [review]

File: src/solana/solanaPlugin.js [hunk 6]
```diff
@@ -0,0 +1,169 @@
...
+  }
+
+  async parseUri(
+    uri: string,
+    currencyCode?: string,
+    customTokens?: EdgeMetaToken[]
+  ): Promise<EdgeParsedUri> {
+    const networks = { [this.pluginId]: true }
+
+    const { parsedUri, edgeParsedUri } = this.parseUriCommon(
```
[the remark this answers is not in the document]
Comment (peachbits) [reply src/solana/solanaPlugin.js:87]: Yes it's standard bip21 https://github.com/solana-labs/solana-pay/blob/master/SPEC.md we'll need special handling for spl-token once we add token support

**Review by paullinator — APPROVED** [review]

**Pull request merged**